### PR TITLE
sensu check cron scheduling support

### LINKED ISSUE #551: Cron syntax scheduling unusable with sensu chef cookbook
## Expected Behavior

- Interval currently defaults to 60 if not specified with the sensu_check resource, however interval should not be present if the check is making use of cron based scheduling

## Current Behavior

- There is currently no resource for cron scheduling, it must be done as an additional attribute
- When specifying the cron as an additional attribute, the chef provider still populates the check JSON with an interval attribute

## Possible Solution
Add a cron resource to sensu_check. Do not require 'interval' if cron attribute is present.

## Steps to Reproduce (for bugs)
1. Create barebones sensu check, i.e.:
````
sensu_check 'batch_job_check' do
  command 'batch_job_check.rb'
  subscribers ['roundrobin:big-data']
  additional(:cron => '0 9 * * 1', :ttl => 612_000, :source => 'big-data')
end
````
2. After chef converge, check is generated with an interval attribute:
````
{
  "checks": {
    "batch_job_check": {
      "command": "batch_job_check.rb",
      "subscribers": [
        "roundrobin:big-data"
      ],
      "interval": 60,
      "cron": "0 9 * * 1",
      "ttl": 612000,
      "source": "big-data"
    }
  }
}

````

## Context
Currently unable to deploy cron scheduled checks using the sensu_check chef resource.

## Your Environment

* Version of this cookbook used: 4.0.0
* Version of Sensu used: 0.28.4
* Version of Chef used: 12.5.1 :(
* Operating System and version (e.g. CentOS 7, Ubuntu 14.04): CentOS 7.3

### PR DESCRIPTION
**Pull request by vbichov**
adds cron configuration to sensu checks
## Description

Added the code relevant to check configuration and some tests

## Motivation and Context

fixes #551
## How Has This Been Tested?

Added relevant test
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: providers/check.rb [hunk 1]
```diff
@@ -29,8 +29,15 @@ def load_current_resource
       timeout
       ttl
       type
+      interval
+      cron
     ]
-  ).merge("interval" => new_resource.interval).merge(new_resource.additional)
+  ).merge(new_resource.additional).tap do |helper|
+    if helper.key?("interval") && helper.key?("cron")
+      Chef::Log.warn("sensu_check[#{new_resource.name}]: cron and interval don't wrok together.")
```
Comment (majormoses) [inline providers/check.rb]: small typo `work`

File: resources/check.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@
 attribute :command, :kind_of => String, :required => true
 attribute :aggregate, :kind_of => [String, TrueClass, FalseClass]
 attribute :aggregates, :kind_of => Array
-attribute :interval, :default => 60
+attribute :interval, :kind_of => Integer
```
Comment (majormoses) [inline resources/check.rb]: This makes this a breaking change and would require lots of people to potentially update their cookbooks. I think this is how we would do it if we had known that `cron` was gonna be a thing when we made the cookbook and decided on a default for interval. I think we want a default of 60 still with a `:kind_of => [Integer, FalseClass` and would require anyone wishing to use `cron` to set `interval` to `false` this prevents a breaking change.

After reaching the end of the PR backwards compatibility is maintained but is less obvious to the user.

File: providers/check.rb [hunk 3]
```diff
@@ -29,8 +29,15 @@ def load_current_resource
       timeout
       ttl
       type
+      interval
+      cron
     ]
-  ).merge("interval" => new_resource.interval).merge(new_resource.additional)
+  ).merge(new_resource.additional).tap do |helper|
+    if helper.key?("interval") && helper.key?("cron")
```
Comment (majormoses) [inline providers/check.rb]: you probably will need to adjust this based on the change to interval, you probably would want to remove the key from the hash if its value is `false` or `!positive?` (which would account for `0` and `negative` values)

File: resources/check.rb [hunk 4]
```diff
@@ -19,12 +19,22 @@
 attribute :timeout, :kind_of => Integer
 attribute :ttl, :kind_of => Integer
 attribute :type, :kind_of => String, :equal_to => %w[standard status metric]
+attribute :cron, :kind_of => String
```
Comment (majormoses) [inline resources/check.rb]: I think it should be `:kind_of => [String, FalseClass]`

File: resources/check.rb [hunk 5]
```diff
@@ -19,12 +19,22 @@
...
   unless name =~ /^[\w\.-]+$/
     raise Chef::Exceptions::ValidationFailed, "Sensu check #{name}: name cannot contain spaces or special characters"
   end
 
+  if !cron.nil? && !interval.nil?
+    raise Chef::Exceptions::ValidationFailed, "Sensu check #{name}: must either define interval, or cron, not both."
+  end
+
+  # preserve existing defalut value for interval
+  if cron.nil? && interval.nil?
```
Comment (majormoses) [inline resources/check.rb]: I see this is where you are preserving backwards compatibility, which I thank you for doing so. I think there is a more straightforward way than this as I have outlined above.

**Comment by majormoses** [comment]
I would say if you have integration tests feel free to remove the unit tests. If they were a library function it would make more sense to write a unit test for it.

**Comment by vbichov** [comment]
Hi,
Fixed the issues - tests pass.

**Comment by majormoses** [comment]
Sorry this fell off my radar. I will work on merging, writing up a changelog entry, and cutting a release.

**Comment by majormoses** [comment]
released: https://supermarket.chef.io/cookbooks/sensu/versions/5.3.0

**Pull request merged**